## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#bc1f7df7-5beb-43bc-8548-34665ddc6b90) |
| mint-dist-erasure.sh | :x: [more...](#8756db50-b6e3-449c-8d39-c6331558ec36) |
| mint-zoned.sh | :x: [more...](#da8cc05d-54c8-4791-83f6-8c4dcd17fa8f) |
<a name="da8cc05d-54c8-4791-83f6-8c4dcd17fa8f"></a>
### 10600-da924a2/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:32270
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d317d94e1596:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 1 minutes and 19 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 34 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 8 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 53 seconds
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
main.testCopyObjectWithVersioning()
	/mint/run/core/minio-go/main.go:1108 +0x2138
main.main()
	/mint/run/core/minio-go/main.go:11285 +0x230
(9/15) Running minio-java tests ... done in 2 minutes and 24 seconds
(10/15) Running minio-js tests ... done in 49 seconds
(11/15) Running minio-py tests ... done in 3 minutes and 47 seconds
(12/15) Running s3cmd tests ... done in 21 seconds
(13/15) Running s3select tests ... done in 7 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="bc1f7df7-5beb-43bc-8548-34665ddc6b90"></a>
### 10600-da924a2/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31160
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d4301fff42d4:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 7 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 2 minutes and 40 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 28 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 42 seconds
(9/15) Running minio-go tests ... FAILED in 2 minutes and 1 seconds
{
  "args": {
    "bucketName": "minio-go-test-f1ahn3mbbi1gpetf",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1089,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 10 minutes and 47 seconds
(10/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 79,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... FAILED in 18 minutes and 22 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-a46c5fb9-d970-41c9-b5b3-9cab52cf4cc4",
    "object_name": "30d3092e-ad01-4a66-9ded-b0a556e6b5cb"
  },
  "message": "Sha-sum mismatch on multi-threaded put and get objects",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 145, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1666, in test_thread_safe\n    raise exceptions[0]\n  File \"/mint/run/core/minio-py/tests.py\", line 1631, in get_object_and_check\n    'Sha-sum mismatch on multi-threaded put and '\nValueError: Sha-sum mismatch on multi-threaded put and get objects\n",
  "duration": 28380
}
(10/15) Running s3cmd tests ... done in 2 minutes and 15 seconds
(11/15) Running s3select tests ... done in 40 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="8756db50-b6e3-449c-8d39-c6331558ec36"></a>
### 10600-da924a2/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:31864
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a80b452c8952:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ja9sqaruvv164a",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 179,
  "error": "AWS S3 key count mismatch",
  "function": "testListObjects",
  "message": "AWS SDK Go listobjects with prefix '' expected 2 key but got 0, []",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 1 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "name": "aws-sdk-php",
  "function": "listObjects ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5f73ea5bac131",
    "Object": "obj1"
  },
  "duration": "329",
  "status": "FAIL",
  "error": "listObject API failed for aws-sdk-php-5f73ea5bac131/obj1"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-f25ccc8cbbcb"
  },
  "duration": 448.62,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-f25ccc8cbbcb', The bucket you tried to delete is not empty",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 8 seconds
{
  "name": "awscli",
  "duration": 3537,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "remove_bucket failed: s3://awscli-mint-test-bucket-28662 An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: The bucket you tried to delete is not empty"
}
(2/15) Running healthcheck tests ... done in 0 seconds
(3/15) Running mc tests ... FAILED in 2 seconds
{
  "name": "mc",
  "duration": "280",
  "function": "test_put_object_with_metadata",
  "status": "FAIL",
  "error": "unable to put object with metadata"
}
(3/15) Running minio-dotnet tests ... done in 36 seconds
(4/15) Running minio-go tests ... FAILED in 2 seconds
{
  "args": {
    "bucketName": "minio-go-test-ij6mbv1rlosgnn2t",
    "isRecursive": true,
    "objectName": "minio-go-test-ij6mbv1rlosgnn2tunique"
  },
  "duration": 296,
  "function": "ListObjects(bucketName, objectName, isRecursive, doneCh)",
  "message": "Could not find existing object minio-go-test-ij6mbv1rlosgnn2tunique",
  "name": "minio-go: testFunctionalV2",
  "status": "FAIL"
}
(4/15) Running minio-java tests ... FAILED in 9 seconds
{
  "name": "minio-java",
  "function": "listObjects()",
  "args": "[bucket]",
  "duration": 233,
  "status": "FAIL",
  "error": "java.lang.Exception: object count; expected=3, got=0 >>> [FunctionalTest.testListObjects(FunctionalTest.java:1429), FunctionalTest.listObjects_test(FunctionalTest.java:1449), FunctionalTest.runTests(FunctionalTest.java:3811), FunctionalTest.main(FunctionalTest.java:4058)]"
}
(4/15) Running minio-js tests ... FAILED in 17 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 9,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(4/15) Running minio-py tests ... FAILED in 58 seconds
{
  "name": "minio-py:test_list_objects_with_prefix",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-2eeaca8f-232a-4d0c-8b28-1274a2c16ce5",
    "object_name": "47d5e2e1-412d-45b1-a0c9-3baa03c96d3e"
  },
  "message": "Listed no of objects (0), does not match the expected no of objects (4)",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 145, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1208, in test_list_objects_with_prefix\n    _test_list_objects_api(bucket_name, no_of_created_files, prefix, True)\n  File \"/mint/run/core/minio-py/tests.py\", line 1162, in _test_list_objects_api\n    \"expected no of objects ({})\").format(no_of_files, expected_no))\nValueError: Listed no of objects (0), does not match the expected no of objects (4)\n",
  "duration": 417
}
(4/15) Running s3cmd tests ... FAILED in 24 seconds
{
  "name": "s3cmd",
  "duration": "12674",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "sync and list differs"
}
(4/15) Running s3select tests ... done in 6 seconds
(5/15) Running security tests ... done in 0 seconds

Executed 5 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10600-da924a2

--- a/cmd/disk-cache-backend.go
+++ b/cmd/disk-cache-backend.go
@@ -194,7 +194,7 @@ func (c *diskCache) diskUsageLow() bool {
 		logger.LogIf(ctx, err)
 		return false
 	}
-	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	usedPercent := (di.Used / di.Total) * 100
 	low := int(usedPercent) < gcStopPct
 	atomic.StoreUint64(&c.stats.UsagePercent, usedPercent)
 	if low {
@@ -218,7 +218,7 @@ func (c *diskCache) diskSpaceAvailable(size int64) bool {
 		logger.Info("diskCache: Received 0 total disk size")
 		return false
 	}
-	usedPercent := float64(di.Total-di.Free) * 100 / float64(di.Total)
+	usedPercent := float64(di.Used) * 100 / float64(di.Total)
 	if usedPercent >= float64(gcTriggerPct) {
 		atomic.StoreInt32(&c.stats.UsageState, 1)
 		c.queueGC()
@@ -226,7 +226,7 @@ func (c *diskCache) diskSpaceAvailable(size int64) bool {
 	atomic.StoreUint64(&c.stats.UsagePercent, uint64(usedPercent))
 
 	// Recalculate percentage with provided size added.
-	usedPercent = float64(di.Total-di.Free+uint64(size)) * 100 / float64(di.Total)
+	usedPercent = float64(di.Used+uint64(size)) * 100 / float64(di.Total)
 
 	return usedPercent < float64(c.quotaPct)
 }

--- a/cmd/erasure-common.go
+++ b/cmd/erasure-common.go
@@ -59,7 +59,7 @@ func (er erasureObjects) getLoadBalancedDisks() []StorageAPI {
 
 	var wg sync.WaitGroup
 	var mu sync.Mutex
-	var newDisks []StorageAPI
+	var newDisks = map[uint64][]StorageAPI{}
 	// Based on the random shuffling return back randomized disks.
 	for _, i := range hashOrder(UTCNow().String(), len(disks)) {
 		i := i
@@ -79,13 +79,24 @@ func (er erasureObjects) getLoadBalancedDisks() []StorageAPI {
 				// - Future: skip busy disks
 				return
 			}
+
 			mu.Lock()
-			newDisks = append(newDisks, disks[i-1])
+			// Capture disks usage wise
+			newDisks[di.Used] = append(newDisks[di.Used], disks[i-1])
 			mu.Unlock()
 		}()
 	}
 	wg.Wait()
-	return newDisks
+
+	var max uint64
+	for k := range newDisks {
+		if k > max {
+			max = k
+		}
+	}
+
+	// Return disks which have maximum disk usage common.
+	return newDisks[max]
 }
 
 // This function does the following check, suppose

--- a/cmd/erasure-zones.go
+++ b/cmd/erasure-zones.go
@@ -662,7 +662,7 @@ func (z *erasureZones) listObjectsNonSlash(ctx context.Context, bucket, prefix, 
 	for _, zone := range z.zones {
 		zonesEntryChs = append(zonesEntryChs,
 			zone.startMergeWalksN(ctx, bucket, prefix, "", true, endWalkCh, zone.listTolerancePerSet, false))
-		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-1)
+		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-2)
 	}
 
 	var objInfos []ObjectInfo
@@ -784,7 +784,7 @@ func (z *erasureZones) listObjectsSplunk(ctx context.Context, bucket, prefix, ma
 		}
 		zonesEntryChs = append(zonesEntryChs, entryChs)
 		zonesEndWalkCh = append(zonesEndWalkCh, endWalkCh)
-		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-1)
+		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-2)
 	}
 
 	entries := mergeZonesEntriesCh(zonesEntryChs, maxKeys, zonesListTolerancePerSet)
@@ -876,7 +876,7 @@ func (z *erasureZones) listObjects(ctx context.Context, bucket, prefix, marker, 
 		}
 		zonesEntryChs = append(zonesEntryChs, entryChs)
 		zonesEndWalkCh = append(zonesEndWalkCh, endWalkCh)
-		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-1)
+		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-2)
 	}
 
 	entries := mergeZonesEntriesCh(zonesEntryChs, maxKeys, zonesListTolerancePerSet)
@@ -1278,7 +1278,7 @@ func (z *erasureZones) listObjectVersions(ctx context.Context, bucket, prefix, m
 		}
 		zonesEntryChs = append(zonesEntryChs, entryChs)
 		zonesEndWalkCh = append(zonesEndWalkCh, endWalkCh)
-		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-1)
+		zonesListTolerancePerSet = append(zonesListTolerancePerSet, zone.listTolerancePerSet-2)
 	}
 
 	entries := mergeZonesEntriesVersionsCh(zonesEntryChs, maxKeys, zonesListTolerancePerSet)

--- a/cmd/fs-v1.go
+++ b/cmd/fs-v1.go
@@ -216,12 +216,11 @@ func (fs *FSObjects) StorageInfo(ctx context.Context, _ bool) (StorageInfo, []er
 	if err != nil {
 		return StorageInfo{}, []error{err}
 	}
-	used := di.Total - di.Free
 	storageInfo := StorageInfo{
 		Disks: []madmin.Disk{
 			{
 				TotalSpace:     di.Total,
-				UsedSpace:      used,
+				UsedSpace:      di.Used,
 				AvailableSpace: di.Free,
 				DrivePath:      fs.fsPath,
 			},

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -319,6 +319,22 @@ func initAllSubsystems(ctx context.Context, newObject ObjectLayer) (err error) {
 		if err != nil {
 			return fmt.Errorf("Unable to list buckets to heal: %w", err)
 		}
+		rquorum := InsufficientReadQuorum{}
+		wquorum := InsufficientWriteQuorum{}
+		for _, bucket := range buckets {
+			if err = newObject.MakeBucketWithLocation(ctx, bucket.Name, BucketOptions{}); err != nil {
+				if errors.As(err, &wquorum) || errors.As(err, &rquorum) {
+					// Return the error upwards for the caller to retry.
+					return fmt.Errorf("Unable to heal bucket: %w", err)
+				}
+				if _, ok := err.(BucketExists); !ok {
+					// ignore any other error and log for investigation.
+					logger.LogIf(ctx, err)
+					continue
+				}
+				// Bucket already exists, nothing that needs to be done.
+			}
+		}
 	} else {
 		buckets, err = newObject.ListBuckets(ctx)
 		if err != nil {

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -479,7 +479,7 @@ func (s *xlStorage) DiskInfo(context.Context) (info DiskInfo, err error) {
 			}
 			dcinfo.Total = di.Total
 			dcinfo.Free = di.Free
-			dcinfo.Used = di.Total - di.Free
+			dcinfo.Used = di.Used
 			dcinfo.FSType = di.FSType
 
 			diskID, err := s.GetDiskID()

--- a/pkg/disk/disk.go
+++ b/pkg/disk/disk.go
@@ -25,10 +25,8 @@ package disk
 type Info struct {
 	Total  uint64
 	Free   uint64
+	Used   uint64
 	Files  uint64
 	Ffree  uint64
 	FSType string
-
-	// Usage is calculated per tenant.
-	Usage uint64
 }

--- a/pkg/disk/stat_bsd.go
+++ b/pkg/disk/stat_bsd.go
@@ -19,6 +19,7 @@
 package disk
 
 import (
+	"fmt"
 	"syscall"
 )
 
@@ -37,5 +38,9 @@ func GetInfo(path string) (info Info, err error) {
 		Ffree:  uint64(s.Ffree),
 		FSType: getFSType(s.Fstypename[:]),
 	}
+	if info.Free > info.Total {
+		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
+	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

--- a/pkg/disk/stat_linux.go
+++ b/pkg/disk/stat_linux.go
@@ -45,5 +45,6 @@ func GetInfo(path string) (info Info, err error) {
 	if info.Free > info.Total {
 		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
 	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

--- a/pkg/disk/stat_linux_32bit.go
+++ b/pkg/disk/stat_linux_32bit.go
@@ -76,5 +76,6 @@ func GetInfo(path string) (info Info, err error) {
 	if info.Free > info.Total {
 		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
 	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

--- a/pkg/disk/stat_linux_s390x.go
+++ b/pkg/disk/stat_linux_s390x.go
@@ -76,5 +76,6 @@ func GetInfo(path string) (info Info, err error) {
 	if info.Free > info.Total {
 		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
 	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

--- a/pkg/disk/stat_netbsd.go
+++ b/pkg/disk/stat_netbsd.go
@@ -19,6 +19,8 @@
 package disk
 
 import (
+	"fmt"
+
 	"golang.org/x/sys/unix"
 )
 
@@ -36,5 +38,9 @@ func GetInfo(path string) (info Info, err error) {
 		Ffree:  uint64(s.Ffree),
 		FSType: string(s.Fstypename[:]),
 	}
+	if info.Free > info.Total {
+		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
+	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

--- a/pkg/disk/stat_openbsd.go
+++ b/pkg/disk/stat_openbsd.go
@@ -19,6 +19,7 @@
 package disk
 
 import (
+	"fmt"
 	"syscall"
 )
 
@@ -37,5 +38,9 @@ func GetInfo(path string) (info Info, err error) {
 		Ffree:  uint64(s.F_ffree),
 		FSType: getFSType(s.F_fstypename[:]),
 	}
+	if info.Free > info.Total {
+		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
+	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

--- a/pkg/disk/stat_solaris.go
+++ b/pkg/disk/stat_solaris.go
@@ -19,6 +19,8 @@
 package disk
 
 import (
+	"fmt"
+
 	"golang.org/x/sys/unix"
 )
 
@@ -36,5 +38,9 @@ func GetInfo(path string) (info Info, err error) {
 		Ffree:  uint64(s.Ffree),
 		FSType: getFSType(s.Fstr[:]),
 	}
+	if info.Free > info.Total {
+		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
+	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

--- a/pkg/disk/stat_windows.go
+++ b/pkg/disk/stat_windows.go
@@ -19,6 +19,7 @@
 package disk
 
 import (
+	"fmt"
 	"os"
 	"syscall"
 	"unsafe"
@@ -90,5 +91,9 @@ func GetInfo(path string) (info Info, err error) {
 	info.Files = uint64(lpTotalNumberOfClusters)
 	info.Ffree = uint64(lpNumberOfFreeClusters)
 
+	if info.Free > info.Total {
+		return info, fmt.Errorf("detected free space (%d) > total disk space (%d), fs corruption at (%s). please run 'fsck'", info.Free, info.Total, path)
+	}
+	info.Used = info.Total - info.Free
 	return info, nil
 }

